tests/model_inheritance/test_abstract_inheritance.py

```diff
@@ -50,14 +55,76 @@ class Meta:
         class Child(ParentA, ParentB):
             pass
 
-        self.assertEqual(Child.check(), [
-            Error(
-                "The field 'name' clashes with the field 'name' from model "
-                "'model_inheritance.child'.",
-                obj=Child._meta.get_field('name'),
-                id='models.E006',
-            ),
-        ])
+        self.assertEqual(Child.check(), [])
+        inherited_field = Child._meta.get_field('name')
+        self.assertTrue(isinstance(inherited_field, models.CharField))
+        self.assertEqual(inherited_field.max_length, 255)
+
+    def test_diamond_shaped_multiple_inheritance_is_depth_first(self):
+        """
+        In contrast to standard Python MRO, resolution of inherited fields is
+        depth-first, rather than breadth-first.
+
+        This is because a copy of the parent field descriptor is placed onto
+        the model class in ModelBase.__new__(), rather than the attribute
+        lookup going via bases. (It only **looks** like inheritance.)
+
+        Here, Child inherits name from Root, rather than ParentB.
+        """
```

Comment:
Would it be worth emphasising that this straying from pythonic expectation is an _undesirable_ behaviour?
In case `ModelBase.__new__()` were to get rewritten in future, and whoever works on it ends up thinking they need to be supporting these testcases?
Orr just writing the testcase as an expected failure would imply this.

Reply:
I don't see it as undesirable. Just a fact about the way models are built. 

We already document that inheritance isn't quite like Python inheritance. This is just part of that. 

It's not likely that rewriting `ModelBase.__new__()` is really on the cards.